### PR TITLE
feat(dynamodb): Inclusive grants for dynamodb

### PR DESCRIPTION
**Pull request by ahammond**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

Rationale: do the right thing even when granting for a table referenced using a naive `fromArn`. Put another way, we don't know that there aren't secondary indexes, but we do know that the user wants to grant read / write access to the "table", which at L2 is less surprising when it means both the actual table and it's associated indexes.

I don't have working tests yet. Ad-hoc testing shows that, as expected, additional index resources in the iam grants do not break functionality for tables with no indexes. 

Fixes: https://github.com/aws/aws-cdk/issues/13703

### CODE REVIEW

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts [hunk 1]
```diff
@@ -900,6 +892,7 @@ abstract class TableBase extends Resource implements ITable {
   /**
    * Adds an IAM policy statement associated with this table to an IAM
    * principal's policy.
+   * NOTE: Always include index resources, whether indexes are known to exist or not.
    * @param grantee The principal (no-op if undefined)
```
Comment (skinny85) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:896]:
```suggestion
   * principal's policy.
   *
   * **Note**: this method always adds permissions to all of the Table's indexes as well,
   *   whether indexes are known to exist or not.
   *
   * @param grantee The principal (no-op if undefined)
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts [hunk 2]
```diff
@@ -909,11 +902,9 @@ abstract class TableBase extends Resource implements ITable {
   ): iam.Grant {
     if (opts.tableActions) {
       const resources = [this.tableArn,
-        Lazy.string({ produce: () => this.hasIndex ? `${this.tableArn}/index/*` : Aws.NO_VALUE }),
```
Comment (skinny85) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:912]: Is the `hasIndex` field used anywhere else in this class? Can we just remove it now?

**Review by skinny85 — COMMENTED** [review]
LGTM! Couple of minor comments/questions.

I also want @RomainMuller to take a look as well.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts (hunk 2 above)
Comment (jogold) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:912]: and drop the `Lazy`s (everything is known at construction time now).

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts [hunk 3]
```diff
@@ -577,11 +577,9 @@ abstract class TableBase extends Resource implements ITable {
       actions,
       resourceArns: [
         this.tableArn,
-        Lazy.string({ produce: () => this.hasIndex ? `${this.tableArn}/index/*` : Aws.NO_VALUE }),
+        `${this.tableArn}/index/*`,
         ...this.regionalArns,
-        ...this.regionalArns.map(arn => Lazy.string({
-          produce: () => this.hasIndex ? `${arn}/index/*` : Aws.NO_VALUE,
-        })),
+        ...this.regionalArns.map(arn => {return `${arn}/index/*`;}),
```
Comment (jogold) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:582]:
```suggestion
        ...this.regionalArns.map(arn => `${arn}/index/*`),
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts (hunk 2 above)
Comment (ahammond) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:912]: I removed hasIndex and the Lazy's.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts (hunk 3 above)
Comment (ahammond) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: Ooh, thanks!

**Comment by ahammond** [comment]
@skinny85 @jogold @RomainMuller can one of you please help me fix the failing UTs? I'm out of my depth.

**Comment by skinny85** [comment]
@ahammond hopefully, the errors should be clear from the message in the tests. Example from `test/dynamodb.test.js`, `grants - "grant" allows adding arbitrary actions associated with this table resource (via testGrant)`.

The test is expecting:

```
@aws-cdk/aws-dynamodb:     None of 1 resources matches resource 'AWS::IAM::Policy' with {
@aws-cdk/aws-dynamodb:       "$objectLike": {
@aws-cdk/aws-dynamodb:         "PolicyDocument": {
@aws-cdk/aws-dynamodb:           "Statement": [
@aws-cdk/aws-dynamodb:             {
@aws-cdk/aws-dynamodb:               "Action": [
@aws-cdk/aws-dynamodb:                 "dynamodb:action1",
@aws-cdk/aws-dynamodb:                 "dynamodb:action2"
@aws-cdk/aws-dynamodb:               ],
@aws-cdk/aws-dynamodb:               "Effect": "Allow",
@aws-cdk/aws-dynamodb:               "Resource": [
@aws-cdk/aws-dynamodb:                 {
@aws-cdk/aws-dynamodb:                   "Fn::GetAtt": [
@aws-cdk/aws-dynamodb:                     "mytable0324D45C",
@aws-cdk/aws-dynamodb:                     "Arn"
@aws-cdk/aws-dynamodb:                   ]
@aws-cdk/aws-dynamodb:                 },
@aws-cdk/aws-dynamodb:                 {
@aws-cdk/aws-dynamodb:                   "Fn::Join": [
@aws-cdk/aws-dynamodb:                     "",
@aws-cdk/aws-dynamodb:                     [
@aws-cdk/aws-dynamodb:                       {
@aws-cdk/aws-dynamodb:                         "Fn::GetAtt": [
@aws-cdk/aws-dynamodb:                           "TableA3D7B5AFA",
@aws-cdk/aws-dynamodb:                           "Arn"
@aws-cdk/aws-dynamodb:                         ]
@aws-cdk/aws-dynamodb:                       },
@aws-cdk/aws-dynamodb:                       "/index/*"
@aws-cdk/aws-dynamodb:                     ]
@aws-cdk/aws-dynamodb:                   ]
@aws-cdk/aws-dynamodb:                 }
@aws-cdk/aws-dynamodb:               ]
@aws-cdk/aws-dynamodb:             }
@aws-cdk/aws-dynamodb:           ],
@aws-cdk/aws-dynamodb:           "Version": "2012-10-17"
@aws-cdk/aws-dynamodb:         },
@aws-cdk/aws-dynamodb:         "PolicyName": "userDefaultPolicy083DF682",
@aws-cdk/aws-dynamodb:         "Users": [
@aws-cdk/aws-dynamodb:           {
@aws-cdk/aws-dynamodb:             "Ref": "user2C2B57AE"
@aws-cdk/aws-dynamodb:           }
@aws-cdk/aws-dynamodb:         ]
@aws-cdk/aws-dynamodb:       }
@aws-cdk/aws-dynamodb:     }.
```

But it found:

```
@aws-cdk/aws-dynamodb:     - Field PolicyDocument mismatch: Field Statement mismatch: Array element 0 mismatch in:
@aws-cdk/aws-dynamodb:         {
@aws-cdk/aws-dynamodb:           "Type": "AWS::IAM::Policy",
@aws-cdk/aws-dynamodb:           "Properties": {
@aws-cdk/aws-dynamodb:             "PolicyDocument": {
@aws-cdk/aws-dynamodb:               "Statement": [
@aws-cdk/aws-dynamodb:                 {
@aws-cdk/aws-dynamodb:                   "Action": [
@aws-cdk/aws-dynamodb:                     "dynamodb:action1",
@aws-cdk/aws-dynamodb:                     "dynamodb:action2"
@aws-cdk/aws-dynamodb:                   ],
@aws-cdk/aws-dynamodb:                   "Effect": "Allow",
@aws-cdk/aws-dynamodb:                   "Resource": [
@aws-cdk/aws-dynamodb:                     {
@aws-cdk/aws-dynamodb:                       "Fn::GetAtt": [
@aws-cdk/aws-dynamodb:                         "mytable0324D45C",
@aws-cdk/aws-dynamodb:                         "Arn"
@aws-cdk/aws-dynamodb:                       ]
@aws-cdk/aws-dynamodb:                     },
@aws-cdk/aws-dynamodb:                     {
@aws-cdk/aws-dynamodb:                       "Fn::Join": [
@aws-cdk/aws-dynamodb:                         "",
@aws-cdk/aws-dynamodb:                         [
@aws-cdk/aws-dynamodb:                           {
@aws-cdk/aws-dynamodb:                             "Fn::GetAtt": [
@aws-cdk/aws-dynamodb:                               "mytable0324D45C",
@aws-cdk/aws-dynamodb:                               "Arn"
@aws-cdk/aws-dynamodb:                             ]
@aws-cdk/aws-dynamodb:                           },
@aws-cdk/aws-dynamodb:                           "/index/*"
@aws-cdk/aws-dynamodb:                         ]
@aws-cdk/aws-dynamodb:                       ]
@aws-cdk/aws-dynamodb:                     }
@aws-cdk/aws-dynamodb:                   ]
@aws-cdk/aws-dynamodb:                 }
@aws-cdk/aws-dynamodb:               ],
@aws-cdk/aws-dynamodb:               "Version": "2012-10-17"
@aws-cdk/aws-dynamodb:             },
@aws-cdk/aws-dynamodb:             "PolicyName": "userDefaultPolicy083DF682",
@aws-cdk/aws-dynamodb:             "Users": [
@aws-cdk/aws-dynamodb:               {
@aws-cdk/aws-dynamodb:                 "Ref": "user2C2B57AE"
@aws-cdk/aws-dynamodb:               }
@aws-cdk/aws-dynamodb:             ]
@aws-cdk/aws-dynamodb:           }
@aws-cdk/aws-dynamodb:         }
```

So [this code](https://github.com/aws/aws-cdk/blob/d82de0518abf4deb2c5b38d8eed9653fa8181cbd/packages/%40aws-cdk/aws-dynamodb/test/dynamodb.test.ts#L2864-L2891) will have to be updated with the new statements.

Similar changes are needed for other tests.

Let me know if this is enough info to fix these @ahammond!

**Comment by skinny85** [comment]
Closing due to staleness - if you're still working on this @ahammond, please re-open, or comment, and I'll re-open it.

**Pull request closed without merging**